src/mpyl/stages/discovery.py

```diff
@@ -87,15 +87,55 @@ def is_dependency_touched(
     return False
 
 
-def is_output_cached(output: Optional[Output], cache_key: str) -> bool:
-    if (
-        output is None
-        or not output.success
-        or output.produced_artifact is None
-        or not output.produced_artifact.hash
-    ):
-        return False
-    return output.produced_artifact.hash == cache_key
+def is_project_cached_for_stage(
+    logger: logging.Logger,
+    project: str,
+    stage: str,
+    output: Optional[Output],
+    cache_key: str,
+) -> bool:
+    if stage == deploy.STAGE_NAME:
+        logger.debug(
+            f"Project {project} will execute stage {stage} again because this stage is never cached"
+        )
+        cached = False
+    elif output is None:
+        logger.debug(
+            f"Project {project} will execute stage {stage} again because there is no previous run"
+        )
+        cached = False
+    elif not output.success:
+        logger.debug(
+            f"Project {project} will execute stage {stage} again because the previous run was not successful"
+        )
+        cached = False
+    elif output.produced_artifact is None:
+        logger.debug(
+            f"Project {project} will execute stage {stage} again because there was no artifact in the previous run"
+        )
+        cached = False
+    elif not output.produced_artifact.hash:
+        logger.debug(
+            f"Project {project} will execute stage {stage} again because there is no cache key in the previous run"
+        )
+        cached = False
+    elif output.produced_artifact.hash != cache_key:
+        logger.debug(
+            f"Project {project} will execute stage {stage} again because its content changed since the previous run"
+        )
+        logger.debug(
+            f"Cached content for previous run: {output.produced_artifact.hash}"
+        )
+        logger.debug(f"Cached content for current run:  {cache_key}")
```

Comment:
Wouldn't saying 'cached content' be confusing when we log the cache key?

Reply:
Fair, I'll find a better wording.